### PR TITLE
Gas optimizations

### PR DESCRIPTION
**Pull request by pixelcircuits**
- Reduced amount of data passed to embedded standards
- Removed embedded intents as their own inherited contract
- Added limit of 256 segments per intent and 32 intents per solution to reduce amount of dynamic data required when processing
- Combined intent standards of different curve types to simplify code
- Reduced the amount of data a curve definition consumes
- Reduced the amount of storage required for Entry Point security
- Removed emitting events for solver's intents

Gas Consumption:
- ERC20 Transfer (fee paid in same token): 138780 (110039 for a batch of 4)
    - reference - 51354 for an ERC20 transfer tx
- ETH Transfer (fee paid in other token): 146371 (117618 for a batch of 4)
    - reference - 21000 for a simple ETH transfer
- ERC20 Swap (fee paid intrinsically): 199716 (114010 for a batch of 4)
    - reference - 117437 for a simple swap to WETH and unwrapped to ETH

### CODE REVIEW

**Comment by pixelcircuits** [comment]
Yikes. Unfortunately forge coverage is now broken due to the `--via-ir` compilation flag. This flag is necessary in order to save gas. I'm not sure how to get around this... it is a [known issue](https://github.com/foundry-rs/foundry/issues/3527) for forge though.

We may have to switch to hardhat for coverage testing as it does not output the same error

**Comment by pixelcircuits** [comment]
So I reworked the `EntryPoint` a bit to get past the "stack too deep" compilation errors so that coverage will still work. `via-ir` is still used during hardhat tests (and future benchmarking) which does save some gas, but performance has still taken about an 800 gas hit. It might be worth merging `_handleIntentSegment` into `_handleIntents` right before deployment to save that 800 gas.

File: src/standards/Erc20Release.sol [hunk 1]
```diff
@@ -10,93 +9,128 @@ import {Strings} from "openzeppelin/utils/Strings.sol";
 import {Erc20ReleaseDelegate} from "./delegates/Erc20ReleaseDelegate.sol";
 import {popFromCalldata} from "./utils/ContextData.sol";
 import {getSegmentWord} from "./utils/SegmentData.sol";
-import {evaluateConstantCurve, encodeConstantCurve, encodeAsUint96} from "./utils/CurveCoder.sol";
+import {evaluateCurve, encodeConstantCurve, encodeComplexCurve} from "./utils/CurveCoder.sol";
 import {IERC20} from "openzeppelin/token/ERC20/IERC20.sol";
 
 /**
- * ERC20 Release Intent Standard core logic
+ * ERC20 Release with Exponential Curve Intent Standard core logic
```
Comment (supiket) [inline src/standards/Erc20Release.sol:16]:
```suggestion
 * ERC20 Release Intent Standard core logic
```

File: src/standards/EthRelease.sol [hunk 2]
```diff
@@ -12,97 +9,122 @@ import {Strings} from "openzeppelin/utils/Strings.sol";
 import {EthReleaseDelegate} from "./delegates/EthReleaseDelegate.sol";
 import {popFromCalldata} from "./utils/ContextData.sol";
 import {getSegmentWord} from "./utils/SegmentData.sol";
-import {evaluateConstantCurve, encodeConstantCurve, encodeAsUint96} from "./utils/CurveCoder.sol";
+import {evaluateCurve, encodeConstantCurve, encodeComplexCurve} from "./utils/CurveCoder.sol";
 
 /**
- * Eth Release Intent Standard core logic
+ * Eth Release with Exponential Curve Intent Standard core logic
```
Comment (supiket) [inline src/standards/EthRelease.sol:15]:
```suggestion
 * Eth Release Intent Standard core logic
```

File: src/standards/EthRelease.sol [hunk 3]
```diff
@@ -12,97 +9,122 @@ import {Strings} from "openzeppelin/utils/Strings.sol";
...
-    function encodeData(bytes32 standardId, int256 amount) external pure returns (bytes memory) {
-        (uint96 adjustedAmount, uint8 amountMult, bool amountNegative) = encodeAsUint96(amount);
-        bytes32 data = encodeConstantCurve(uint96(adjustedAmount), amountMult, amountNegative, false);
-        return abi.encodePacked(standardId, bytes14(data));
     }
 }
 
 /**
- * Eth Release Intent Standard that can be deployed and registered to the entry point
+ * Eth Release with Exponential Curve Intent Standard that can be deployed and registered to the entry point
```
Comment (supiket) [inline src/standards/EthRelease.sol:59]:
```suggestion
 * Eth Release Intent Standard that can be deployed and registered to the entry point
```

File: src/standards/EthRequire.sol [hunk 4]
```diff
@@ -1,58 +1,49 @@
...
 import {pop} from "./utils/ContextData.sol";
 import {getSegmentWord} from "./utils/SegmentData.sol";
-import {
-    evaluateConstantCurve, encodeConstantCurve, isConstantCurveRelative, encodeAsUint96
-} from "./utils/CurveCoder.sol";
+import {evaluateCurve, encodeConstantCurve, encodeComplexCurve, isCurveRelative} from "./utils/CurveCoder.sol";
 
 /**
- * Eth Require Intent Standard core logic
+ * Eth Require with Exponential Curve Intent Standard core logic
```
Comment (supiket) [inline src/standards/EthRequire.sol:13]:
```suggestion
 * Eth Require Intent Standard core logic
```

File: src/standards/EthRequire.sol [hunk 5]
```diff
@@ -77,47 +68,75 @@ abstract contract BaseEthRequire is BaseIntentStandard {
...
-    function encodeData(bytes32 standardId, int256 amount, bool isRelative) external pure returns (bytes memory) {
-        (uint96 adjustedAmount, uint8 amountMult, bool amountNegative) = encodeAsUint96(amount);
-        bytes32 data = encodeConstantCurve(uint96(adjustedAmount), amountMult, amountNegative, isRelative);
-        return abi.encodePacked(standardId, bytes14(data));
-    }
 }
 
 /**
- * Eth Release Intent Standard that can be deployed and registered to the entry point
+ * Eth Require with Exponential Curve Intent Standard that can be deployed and registered to the entry point
```
Comment (supiket) [inline src/standards/EthRequire.sol:74]:
```suggestion
 * Eth Require Intent Standard that can be deployed and registered to the entry point
```

File: test/foundry/scenarios/TokenSwaps.sol [hunk 6]
```diff
@@ -98,82 +102,89 @@ contract TokenSwaps is ScenarioTestEnvironment {
...
-        vm.assume(0 < releaseDuration);
-        vm.assume(0 < releaseAt);
-        vm.assume(releaseAt < releaseDuration);
-        vm.assume(ethRequireAmount < erc20ReleaseAmount);
-        vm.assume(erc20ReleaseAmount < _accountInitialERC20Balance);
+    function test_constantExpectation()
+        //uint72 erc20ReleaseAmount,
+        //uint72 ethRequireAmount,
+        //uint16 releaseDuration,
+        //uint16 releaseAt
```
Comment (supiket) [inline test/foundry/scenarios/TokenSwaps.sol:146]:
```suggestion
```

File: test/foundry/scenarios/TokenSwaps.sol [hunk 7]
```diff
@@ -98,82 +102,89 @@ contract TokenSwaps is ScenarioTestEnvironment {
...
+        public
+    {
+        //vm.assume(0 < erc20ReleaseAmount);
+        //vm.assume(0 < ethRequireAmount);
+        //vm.assume(0 < releaseDuration);
+        //vm.assume(0 < releaseAt);
+        //vm.assume(releaseAt < releaseDuration);
+        //vm.assume(ethRequireAmount < erc20ReleaseAmount);
+        //vm.assume(erc20ReleaseAmount < _accountInitialERC20Balance);
```
Comment (supiket) [inline test/foundry/scenarios/TokenSwaps.sol:155]:
```suggestion
```

Comment (pixelcircuits) [reply test/foundry/scenarios/TokenSwaps.sol:155]: forgot to turn fuzzing back on when I was debugging

**Review by supiket — APPROVED** [review]

**Pull request merged**